File: app/src/main/java/com/jwu5/giphyapp/giphyhome/HomeFragment.java
package com.jwu5.giphyapp.giphyhome;

import android.content.res.Configuration;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SearchView;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.jwu5.giphyapp.adapters.GiphyRecyclerViewAdapter;
import com.jwu5.giphyapp.R;
import com.jwu5.giphyapp.network.model.GiphyModel;

import java.util.ArrayList;
public class HomeFragment extends Fragment implements HomeView{
    public static final String TAB_NAME = "Home";
    private static final int LIMIT = 12;

    private RecyclerView mGiphyRecyclerView;
    private GridLayoutManager mGridLayoutManager;
    private GiphyRecyclerViewAdapter mGiphyRecyclerViewAdapter;

    private String mQuery = null;
    private SearchView mSearchView;

    private boolean isLoading = false;
    private boolean trending = true;
    private int mOrientation;

    private HomePresenter mHomePresenter;

    public static HomeFragment newInstance() {
        return new HomeFragment();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);
        setHasOptionsMenu(true);

        mOrientation = getActivity().getResources().getConfiguration().orientation;
        mHomePresenter = new HomePresenter(this);
        mGiphyRecyclerViewAdapter = new GiphyRecyclerViewAdapter(getActivity(), TAB_NAME, mOrientation);
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        mOrientation = newConfig.orientation;
        mHomePresenter.updateUI(mGiphyRecyclerViewAdapter.getItems(), mGridLayoutManager.findFirstVisibleItemPosition());
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_giphy_view, container, false);

        mGiphyRecyclerView = (RecyclerView)v.findViewById(R.id.fragment_giphy_view_recycler_view);
        mGridLayoutManager = new GridLayoutManager(getActivity(), 2);
        mGiphyRecyclerView.setLayoutManager(mGridLayoutManager);
        mGiphyRecyclerView.setAdapter(mGiphyRecyclerViewAdapter);

        if (trending) {
            mHomePresenter.getTrendingList(LIMIT, 0);
        }

        mGiphyRecyclerView.addOnScrollListener(new PaginationScrollListener(mGridLayoutManager) {
            @Override
            public void onLoadMore(int totalItemCount) {
                if (!isLoading) {
                    if(trending) {
                        mHomePresenter.getTrendingList(LIMIT, totalItemCount);
                    } else {
                        mHomePresenter.getSearchedList(mQuery, LIMIT, totalItemCount);
                    }
                }
            }
        });
        return v;
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater menuInflater) {
        super.onCreateOptionsMenu(menu, menuInflater);
        menuInflater.inflate(R.menu.fragment_giphy_view, menu);
        MenuItem searchItem = menu.findItem(R.id.menu_item_search);
        mSearchView = (SearchView) searchItem.getActionView();

        mSearchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String s) {
                String result = s.trim().replaceAll(" ", "-").toLowerCase();
                mQuery = s;
                mGiphyRecyclerViewAdapter.removeAll();
                mHomePresenter.getSearchedList(result, LIMIT, 0);
                mSearchView.clearFocus();
                return true;
            }
            @Override
            public boolean onQueryTextChange(String s) {
                return false;
            }
        });
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch(item.getItemId()) {
            case R.id.menu_item_clear:
                mSearchView.setQuery("", false);
                mSearchView.setIconified(true);
                mHomePresenter.getTrendingList(LIMIT, 0);
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public void showTrendingList(ArrayList<GiphyModel> items) {
        if(!trending) {
            mGiphyRecyclerViewAdapter.setItemList(items);
            trending = true;
            return;
        }
        mGiphyRecyclerViewAdapter.addItems(items);
    }

    @Override
    public void showSearchList(ArrayList<GiphyModel> items) {
        if(trending) {
            mGiphyRecyclerViewAdapter.setItemList(items);
            trending = false;
            return;
        }
        mGiphyRecyclerViewAdapter.addItems(items);
    }

    @Override
    public void setLoading() {
        isLoading = true;
    }

    @Override
    public void setLoadingComplete() {
        isLoading = false;
    }

    @Override
    public void onResume() {
        super.onResume();
        mGiphyRecyclerViewAdapter.notifyDataSetChanged();
    }

    @Override
    public void updateAdapter(ArrayList<GiphyModel> items, int position) {
        mGiphyRecyclerViewAdapter = new GiphyRecyclerViewAdapter(getActivity(), TAB_NAME, mOrientation);
        mGiphyRecyclerViewAdapter.setItemList(items);
        mGiphyRecyclerView.setAdapter(mGiphyRecyclerViewAdapter);
        mGiphyRecyclerView.scrollToPosition(position);
    }
}

File: app/src/main/java/com/jwu5/giphyapp/giphyhome/HomeFragment.kt
package com.jwu5.giphyapp.giphyhome

import android.content.res.Configuration
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.SearchView
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup

import com.jwu5.giphyapp.adapters.GiphyRecyclerViewAdapter
import com.jwu5.giphyapp.R
import com.jwu5.giphyapp.network.model.GiphyModel

import java.util.ArrayList

class HomeFragment : Fragment(), HomeView {

    private var mGiphyRecyclerView: RecyclerView? = null
    private var mGridLayoutManager: GridLayoutManager? = null
    private var mGiphyRecyclerViewAdapter: GiphyRecyclerViewAdapter? = null

    private var mQuery: String? = null
    private var mSearchView: SearchView? = null

    private var isLoading = false
    private var trending = true
    private var mOrientation: Int = 0

    private var mHomePresenter: HomePresenter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        retainInstance = true
        setHasOptionsMenu(true)

        mOrientation = activity.resources.configuration.orientation
        mHomePresenter = HomePresenter(this)
        mGiphyRecyclerViewAdapter = GiphyRecyclerViewAdapter(activity, TAB_NAME, mOrientation)
    }

    override fun onConfigurationChanged(newConfig: Configuration?) {
        mOrientation = newConfig!!.orientation
        mHomePresenter!!.updateUI(mGiphyRecyclerViewAdapter!!.items, mGridLayoutManager!!.findFirstVisibleItemPosition())
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val v = inflater!!.inflate(R.layout.fragment_giphy_view, container, false)

        mGiphyRecyclerView = v.findViewById(R.id.fragment_giphy_view_recycler_view) as RecyclerView
        mGridLayoutManager = GridLayoutManager(activity, 2)
        mGiphyRecyclerView!!.layoutManager = mGridLayoutManager
        mGiphyRecyclerView!!.adapter = mGiphyRecyclerViewAdapter

        if (trending) {
            mHomePresenter!!.getTrendingList(LIMIT, 0)
        }

        mGiphyRecyclerView!!.addOnScrollListener(object : PaginationScrollListener(mGridLayoutManager!!) {
            override fun onLoadMore(totalItemCount: Int) {
                if (!isLoading) {
                    if (trending) {
                        mHomePresenter!!.getTrendingList(LIMIT, totalItemCount)
                    } else {
                        mHomePresenter!!.getSearchedList(mQuery, LIMIT, totalItemCount)
                    }
                }
            }
        })
        return v
    }

    override fun onCreateOptionsMenu(menu: Menu?, menuInflater: MenuInflater?) {
        super.onCreateOptionsMenu(menu, menuInflater)
        menuInflater!!.inflate(R.menu.fragment_giphy_view, menu)
        val searchItem = menu!!.findItem(R.id.menu_item_search)
        mSearchView = searchItem.actionView as SearchView

        mSearchView!!.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(s: String): Boolean {
                val result = s.trim { it <= ' ' }.replace(" ".toRegex(), "-").toLowerCase()
                mQuery = s
                mGiphyRecyclerViewAdapter!!.removeAll()
                mHomePresenter!!.getSearchedList(result, LIMIT, 0)
                mSearchView!!.clearFocus()
                return true
            }

            override fun onQueryTextChange(s: String): Boolean {
                return false
            }
        })
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item!!.itemId) {
            R.id.menu_item_clear -> {
                mSearchView!!.setQuery("", false)
                mSearchView!!.isIconified = true
                mHomePresenter!!.getTrendingList(LIMIT, 0)
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    override fun showTrendingList(items: ArrayList<GiphyModel>?) {
        if (!trending) {
            mGiphyRecyclerViewAdapter!!.setItemList(items)
            trending = true
            return
        }
        mGiphyRecyclerViewAdapter!!.addItems(items)
    }

    override fun showSearchList(items: ArrayList<GiphyModel>?) {
        if (trending) {
            mGiphyRecyclerViewAdapter!!.setItemList(items)
            trending = false
            return
        }
        mGiphyRecyclerViewAdapter!!.addItems(items)
    }

    override fun setLoading() {
        isLoading = true
    }

    override fun setLoadingComplete() {
        isLoading = false
    }

    override fun onResume() {
        super.onResume()
        mGiphyRecyclerViewAdapter!!.notifyDataSetChanged()
    }

    override fun updateAdapter(items: ArrayList<GiphyModel>?, position: Int) {
        mGiphyRecyclerViewAdapter = GiphyRecyclerViewAdapter(activity, TAB_NAME, mOrientation)
        mGiphyRecyclerViewAdapter!!.setItemList(items)
        mGiphyRecyclerView!!.adapter = mGiphyRecyclerViewAdapter
        mGiphyRecyclerView!!.scrollToPosition(position)
    }

    companion object {
        val TAB_NAME = "Home"
        private val LIMIT = 12

        fun newInstance(): HomeFragment {
            return HomeFragment()
        }
    }
}
